### PR TITLE
Add displaying joke

### PR DESCRIPTION
**Pull request by Szulerinio**

### CODE REVIEW

File: src/app/app.component.html [hunk 1]
```diff
@@ -1,5 +1,16 @@
...
+      *ngIf="isJokeLoading; then spinnerTemplate; else jokeTemplate"
+    >
+    </ng-container>
+    <ng-template #spinnerTemplate> <p>loading...</p></ng-template>
+    <ng-template #jokeTemplate>
+      <p class="joke" *ngIf="!isJokeError; else errorTemplate">{{ joke }}</p>
+      <ng-template #errorTemplate
+        ><p class="loadError">An error has occurred</p>
+      </ng-template>
+    </ng-template>
```
Comment (Szulerinio) [inline src/app/app.component.html:13]: Should I go with just 3 times *ngIf ?

File: src/app/app.component.html [hunk 2]
```diff
@@ -1,5 +1,16 @@
...
   <app-card>
+    <ng-container
+      *ngIf="isJokeLoading; then spinnerTemplate; else jokeTemplate"
+    >
+    </ng-container>
+    <ng-template #spinnerTemplate> <p>loading...</p></ng-template>
+    <ng-template #jokeTemplate>
+      <p class="joke" *ngIf="!isJokeError; else errorTemplate">{{ joke }}</p>
+      <ng-template #errorTemplate
+        ><p class="loadError">An error has occurred</p>
```
Comment (radKoz) [inline src/app/app.component.html:11]: add texts to translation file

File: src/app/app.component.html (hunk 1 above)
Comment (radKoz) [reply src/app/app.component.html:13]: what do you mean?

File: src/app/app.component.scss [hunk 3]
```diff
@@ -6,3 +7,21 @@
...
+  color: $dark;
+  font-size: 1.8rem;
+  margin-bottom: 3rem;
+  overflow-wrap: break-word;
+  &::before,
+  &::after {
+    content: '"';
+  }
+}
+.loadError {
```
Comment (radKoz) [inline src/app/app.component.scss:23]: use kebab-case in class names in scss

File: src/app/services/joke.service.ts [hunk 4]
```diff
@@ -1,8 +1,55 @@
...
   providedIn: 'root',
 })
 export class JokeService {
-  constructor() {}
+  public getJoke(name: string, categories: string[]): Observable<string> {
+    //this "salt" is workaround for backend issues with handling name containing words "chuck" and "norris
+    const chuckUpperCaseSalt = '—æ²¼';
+    const chuckLowerCaseSalt = '—æ²½';
+    const norrisUpperCaseSalt = '—æ³¼';
+    const norrisLowerCaseSalt = '—æ³½';
```
Comment (radKoz) [inline src/app/services/joke.service.ts:16]: weird stuff, what are these backend issues you mentioned?

File: src/app/services/joke.service.ts [hunk 5]
```diff
@@ -1,8 +1,55 @@
...
+        })
+      );
+  }
+  public getCategories(): Observable<string[]> {
+    return this.http.get<string[]>(
+      environment.apiURL + Constants.API.CATEGORIES
+    );
+  }
+
+  constructor(private http: HttpClient) {}
```
Comment (radKoz) [inline src/app/services/joke.service.ts:54]: move constructor above class methods

File: src/app/app.component.html (hunk 1 above)
Comment (Szulerinio) [reply src/app/app.component.html:13]:
```
  <ng-container
      *ngIf="isJokeLoading; then spinnerTemplate; else jokeTemplate"
    >
    </ng-container>
    <ng-template #spinnerTemplate> <p>loading...</p></ng-template>
    <ng-template #jokeTemplate>
      <p class="joke" *ngIf="!isJokeError; else errorTemplate">{{ joke }}</p>
      <ng-template #errorTemplate
        ><p class="loadError">An error has occurred</p>
      </ng-template>
    </ng-template>

```
gives the same result as
```
    <p *ngIf="isJokeLoading">loading...</p>
    <p class="joke" *ngIf="!isJokeError && !isJokeLoading">{{ joke }}</p>
    <p class="loadError" *ngIf="isJokeError && !isJokeLoading">
      An error has occurred
    </p>
``` 
andI wonder which approach is better
second one looks simpler but is more like 3 ifs and not if else, so I guess it's worse when it comes to performance

File: src/app/services/joke.service.ts (hunk 4 above)
Comment (Szulerinio) [reply src/app/services/joke.service.ts:16]: When in name parameter case insensitive "chuck" or "norris" is present, then response has weird names inside. 
It looks like backend is trying to replace original "chuck norris" in the joke with our passed lets say "Chuck Norris the second" and then again checks for "Chuck" and "Norris" on already changed joke so it changes "Chuck Norris the second" to 
"Chuck Norris the second Norris the second"
and then when looking for "Norris" to 
"Chuck Chuck Norris the second the second Chuck Norris the second the second" 

example: [https://api.chucknorris.io/jokes/random?name=Chuck%20Norris%20the%20second](https://api.chucknorris.io/jokes/random?name=Chuck%20Norris%20the%20second)

I don't know how i would go around this problem in other way than this, or fetching original joke and replacing "Chuck Norris" by myself 

I decided that "—æ³¼" is  rare enough, and noone will put it in the name field

File: src/app/app.component.html (hunk 1 above)
Comment (radKoz) [reply src/app/app.component.html:15]: its better to use if else rather than multiple ifs so I would go with 1st solution

**Review by radKoz — APPROVED** [review]

**Pull request merged**